UniversalAlgebra/Structures/AsRecordsHoms.lagda

```diff
@@ -0,0 +1,223 @@
+ is-hom h = is-hom-rel h × is-hom-op h
+
+ hom : Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ hom = Σ[ h ∈ ((carrier 𝑨) → (carrier 𝑩)) ] is-hom h
+
+-- EXAMPLE.  The special case when 𝑅 = ∅ (i.e., purely algebraic structures)
+module _ {𝐹 : signature}(𝑨 : structure {α} {ρ₀} Sig∅ 𝐹)(𝑩 : structure {β} {ρ₁} Sig∅ 𝐹) where
+
+ -- The type of homomorphisms from one algebraic structure to another.
+ hom-alg : Type (α ⊔ β)
+ hom-alg = Σ[ h ∈ ((carrier 𝑨) → (carrier 𝑩)) ] is-hom-op 𝑨 𝑩 h
+
+
+module _ {𝑅 𝐹 : signature} (𝑨 : structure {α} {ρ₀} 𝑅 𝐹){𝑩 : structure {β} {ρ₁} 𝑅 𝐹}(𝑪 : structure {γ} {ρ₂} 𝑅 𝐹) where
+
+ ∘-is-hom-rel : {f : (carrier 𝑨) → (carrier 𝑩)}{g : (carrier 𝑩) → (carrier 𝑪)}
+  →             is-hom-rel 𝑨 𝑩 f → is-hom-rel 𝑩 𝑪 g → is-hom-rel 𝑨 𝑪 (g ∘ f)
+ ∘-is-hom-rel {f}{g} fhr ghr R a = λ z → ghr R (λ z₁ → f (a z₁)) (fhr R a z)
+
+ ∘-is-hom-op : {f : (carrier 𝑨) → (carrier 𝑩)}{g : (carrier 𝑩) → (carrier 𝑪)}
+  →            is-hom-op 𝑨 𝑩 f → is-hom-op 𝑩 𝑪 g → is-hom-op 𝑨 𝑪 (g ∘ f)
+ ∘-is-hom-op {f}{g} fho gho 𝑓 a = cong g (fho 𝑓 a) ∙ gho 𝑓 (f ∘ a)
+
+ ∘-is-hom : {f : (carrier 𝑨) → (carrier 𝑩)}{g : (carrier 𝑩) → (carrier 𝑪)}
+  →         is-hom 𝑨 𝑩 f → is-hom 𝑩 𝑪 g → is-hom 𝑨 𝑪 (g ∘ f)
+ ∘-is-hom {f} {g} fhro ghro = ihr , iho
+  where
+  ihr : is-hom-rel 𝑨 𝑪 (g ∘ f)
+  ihr = ∘-is-hom-rel {f}{g} (fst fhro) (fst ghro)
+
+  iho : is-hom-op 𝑨 𝑪 (g ∘ f)
+  iho = ∘-is-hom-op {f}{g} (snd fhro) (snd ghro)
+
+ ∘-hom : hom 𝑨 𝑩  →  hom 𝑩 𝑪  →  hom 𝑨 𝑪
+ ∘-hom (f , fh) (g , gh) = g ∘ f , ∘-is-hom {f}{g} fh gh
+
+
+module _ {𝑅 𝐹 : signature} where -- (𝑨 : structure {α} 𝑅 {ρ₀} 𝐹){𝑩 : structure {β} 𝑅 {ρ₁} 𝐹}(𝑪 : structure {γ} 𝑅 {ρ₂} 𝐹) where
+ 𝒾𝒹 : (𝑨 : structure {α} {ρ} 𝑅 𝐹) → hom 𝑨 𝑨
+ 𝒾𝒹 _ = id , (λ R a z → z)  , (λ f a → refl)  -- (λ R a → refl)
+
+module _ {𝑅 𝐹 : signature} (𝑨 : structure {α} {ρ₀} 𝑅 𝐹)(𝑩 : structure {β} {ρ₁} 𝑅 𝐹) where
+
+ is-mon : ((carrier 𝑨) → (carrier 𝑩)) → Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ is-mon g = is-hom 𝑨 𝑩 g × IsInjective g
+
+ mon : Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ mon = Σ[ g ∈ ((carrier 𝑨) → (carrier 𝑩)) ] is-mon g
+
+ mon→hom : mon → hom 𝑨 𝑩
+ mon→hom ϕ = (fst ϕ) , fst (snd ϕ )
+
+
+ is-epi : ((carrier 𝑨) → (carrier 𝑩)) → Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ is-epi g = is-hom 𝑨 𝑩 g × IsSurjective g
+
+ epi : Type (α ⊔ ρ₀ ⊔ β ⊔ ρ₁)
+ epi = Σ[ g ∈ ((carrier 𝑨) → (carrier 𝑩)) ] is-epi g
+
+ epi→hom : epi → hom 𝑨 𝑩
+ epi→hom ϕ = (fst ϕ) , fst (snd ϕ)
+
+module _ {𝑅 𝐹 : signature} where
+ open Lift
+
+ 𝓁𝒾𝒻𝓉 : {𝑨 : structure {α} {ρ} 𝑅 𝐹} → hom 𝑨 (Lift-structure 𝑨 β)
+ 𝓁𝒾𝒻𝓉 = lift , (λ R a x → x) , λ f a → refl
+
+ 𝓁ℴ𝓌ℯ𝓇 : {α β : Level}{𝑨 : structure {α} {ρ} 𝑅 𝐹} → hom (Lift-structure 𝑨 β) 𝑨
+ 𝓁ℴ𝓌ℯ𝓇 = lower , (λ R a x → x) , (λ f a → refl)
+
+-- Kernels of homomorphisms
+
+
+open ≡-Reasoning
+module _ {𝑅 𝐹 : signature} {wd : swelldef ℓ₀ β}{𝑨 : structure {α} {β ⊔ ρ₀} 𝑅 𝐹}{𝑩 : structure {β} {ρ₁} 𝑅 𝐹} where
+
+ homker-comp : (h : hom 𝑨 𝑩) → compatible 𝑨 (ker (fst h))
+ homker-comp h f {u}{v} kuv = ((fst h) (((op 𝑨)f) u))  ≡⟨(snd (snd h)) f u ⟩
```

Comment:
with well-chosen record names, this derivation would not look quite so hideous...

Reply:
Sounds good. Maybe I'll try that, or you could if you get to it before me.

Reply:
I'm working on a POPL paper, which is kind of swallowing all of my Agda time for the next 2.5 weeks. [And I don't even have enough of that, as I've got quite a crew of students who need supervision]

Reply:
fixed